Styles Screen: Add link to global styles revisions

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes https://github.com/WordPress/gutenberg/issues/50429

Add a revisions button to the Styles screen in the site editor's browse mode. Clicking the button exits browse mode, and opens up the site editor proper, with the Revisions screen open.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

As mentioned in #50429, it's helpful to provide quick access to the revisions for global styles.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Add a button to the footer area for the global styles screen in browse mode, displaying the last modified date of the latest revision, the human time, and a revisions icon
* If there are no revisions, or if the revisions haven't loaded, don't display a button at all
* Clicking the button switches to the edit mode for the canvas, and opens the global styles revisions container
* Because we can't update navigation for the global styles sidebar within the editor from outside its navigation provider, add a `useEffect` so that switching to the global styles revisions editor canvas container causes the path to be updated to navigate to the revisions screen. This follows similar behaviour introduced in #49350 by @ntsekouras for linking to the block type panel. I tried a few different approaches, and this seemed the simplest, but happy for any feedback if folks think there are any issues with it!

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. In a blocks-based theme with global styles revisions, open up the site editor and navigate to Styles in browse mode
2. Notice that there should now be a button in the footer area beneath the style variations for selecting revisions
3. Click the button, and the Revisions view in the site editor should open up
4. Click the W icon to go back to the Styles screen
5. Click the editor canvas, and it should open up, but the Revisions screen should not be open
6. Double-check that opening/closing the Revisions screen otherwise works as before

## Questions for the reviewer

* The button design doesn't _quite_ match #50429 — I tried to keep things simple and re-use the footer area of the screen. Please let me know if the button styling should be tweaked. For example, the text elements are spaced via `space-between` but I wasn't sure if more specific spacing was required.
* The text says `Last modified` irrespective of the saved status of the current revision — so if changes have been made but not saved, the text is the same. Is this okay? We already have `1 unsaved change` next to the Save button when there are unsaved changes, so I thought it could be redundant to swap out `Last modified` for `unsaved` 🤔 

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

Check that the button on the Styles screen can be tabbed to. Is the label `Revisions` enough for this button?

## Screenshots or screencast <!-- if applicable -->

The Styles screen — note the addition of the `Last modified` button / row at the bottom left:

![image](https://github.com/WordPress/gutenberg/assets/14988353/afcb9b41-d24a-4de9-87c6-85d7f3b5e458)

https://github.com/WordPress/gutenberg/assets/14988353/6ee1d676-f6e6-4805-9ef5-1f927e28e566

andrewserong committed:
Styles Screen: Add link to global styles revisions

andrewserong committed:
Add button label

ramonjd commented on packages/edit-site/src/components/global-styles/ui.js:
```diff
@@ -237,6 +237,24 @@ function GlobalStylesBlockLink() {
 	}, [ selectedBlockClientId, selectedBlockName, blockHasGlobalStyles ] );
 }
 
+function GlobalStylesRevisionLink() {
```
Is common pattern, e.g., a "hook component"? 😄 

I think I see why it's needed - calling the hook in `GlobalStylesUI()` doesn't seem to work. 

Maybe return `null` for the purists? I don't know 🤷

andrewserong replied:
I originally wrote it as a hook, and then realised it needed to be a component so that it can be executed from within the provider. I then copied the component directly above this one, and added a `return null`, but then removed the `return null` when I saw that the above component doesn't have it.

I personally slightly prefer the explicit `return null`, as it's clearer that it really is a component that doesn't render anything. Happy to add it back in.

ramonjd replied:
Thanks for the explanation! Either way is fine I reckon. It's working as intended!

ramonjd commented on packages/edit-site/src/components/global-styles/ui.js:
```diff
@@ -237,6 +237,24 @@ function GlobalStylesBlockLink() {
+
+	// If the user is in the global styles editor canvas container view, redirect
+	// them to the revisions screen. This effectively allows deep linking to the
+	// revisions screen from outside of the global styles navigation provider.
```
```suggestion
	// revisions screen from outside the global styles navigation provider.
```

Sorry!

andrewserong replied:
No apology necessary, I always love a good copy edit!

ramonjd commented on packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js:
```diff
@@ -69,15 +77,67 @@ function SidebarNavigationScreenGlobalStylesContent() {
+
+	return (
+		<VStack>
+			<SidebarNavigationItem
```
Do you reckon this should have a focus state like the style variation buttons (blue border)?
<img width="335" alt="Screenshot 2023-06-01 at 3 29 08 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/d418dbfa-c77f-4b26-bfa7-a94ebae97354">

andrewserong replied:
Good idea! I like the idea of the blue outline... it might be worth looking at it in a separate PR, though, as it'd apply to all `SidebarNavigationItem`s?

ramonjd replied:
> it might be worth looking at it in a separate PR, though, as it'd apply to all SidebarNavigationItems?

Good call

ramonjd commented on packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js:
```diff
@@ -69,15 +77,67 @@ function SidebarNavigationScreenGlobalStylesContent() {
+	}, [] );
+
+	const hasRevisions = revisionsCount >= 2;
+	const { modified } = revisions?.[ 0 ] || {};
```
Why `{}`?

andrewserong replied:
At this point in the function, we don't know yet if `revisions` is loaded, so without a fallback `{}` I think the destructuring could fail. Here's a fairly artificial example from the console:

<img width="707" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/f94e172e-fef8-4ff1-a8d6-775348f0a7bb">

ramonjd replied:
Oh, my bad for skim reading. Thanks!

ramonjd replied:
I guess `revisions?.[ 0 ]?.modified` might also work? No need to change though, just musing 😄

andrewserong replied:
> I guess revisions?.[ 0 ]?.modified might also work?

Yep, `const modified = revisions?.[ 0 ]?.modified` would also work. I always enjoy these little code style discussions, it's quite illuminating. For me, if I'd started out just wanting to grab the `modified` property, I probably would have used `revisions?.[ 0 ]?.modified`, but while hacking around I was destructuring a few different props, and then whittled it back down to just `modified` so kept the destructuring around, even though it isn't really needed.

Funny how these things hang around! I might swap it out when I do another pass 🙂

ramonjd commented on packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js:
```diff
@@ -95,11 +155,14 @@ export default function SidebarNavigationScreenGlobalStyles() {
-		] );
+	const openGlobalStyles = useCallback(
+		async () =>
+			Promise.all( [
```
Nice 😉

ramonjd reviewed (commented):
This is working as described for me! Checked in Chrome, FF and Safari.

![2023-06-01 15 26 09](https://github.com/WordPress/gutenberg/assets/6458278/e82e952b-450f-4d8b-b608-54c2a8faf610)


Even redirects me to `/revisions` when I click while the site editor loading wheel is still displayed 😄 

Just left a few questions, but it LGTM for now

andrewserong committed:
Update packages/edit-site/src/components/global-styles/ui.js

Co-authored-by: Ramon <ramonjd@users.noreply.github.com>

andrewserong commented:
> Just left a few questions, but it LGTM for now

Thanks for testing, @ramonjd! I was a little stumped with this one until figuring out the `useEffect` in the fake component to handle the redirect. Glad that it's working well in testing — sounds like the next steps will most likely be some design tweaks to make sure it's all looking correct 🙂

tellthemachines commented on packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js:
```diff
@@ -69,15 +77,67 @@ function SidebarNavigationScreenGlobalStylesContent() {
-			onInput={ noop }
-		>
-			<div className="edit-site-sidebar-navigation-screen-global-styles__content">
+		<>
```
Do we need the fragment here?

andrewserong replied:
Oh, we don't! I originally had the extra components in this PR added in the `content` section, but have since moved them to the footer. I've removed the extra fragment in https://github.com/WordPress/gutenberg/pull/51149/commits/ae88d0ba5f9d71b436da3ebe778a97a77397d8e7

tellthemachines reviewed (commented):
Code LGTM! Just a minor question below. 

Also, in testing I noticed that if there are unsaved changes, the revision timestamp doesn't seem to move on from "a few seconds ago", though I don't think that's related to these changes.

Are we waiting on design review? Happy to approve otherwise 😄

jasmussen commented:
Oh that's cool work. GIF showing the revisions button under Styles:

![revisions](https://github.com/WordPress/gutenberg/assets/1204802/d39dad9c-0722-44e0-9683-2bb3f3c394aa)

I would love @jameskoster (back next week) and @SaxonF feedback if either has time, as I know they've worked a lot on these pieces, and are a better canonical source for design feedback. 

But from my angle, a few things:

<img width="437" alt="Screenshot 2023-06-01 at 10 34 59" src="https://github.com/WordPress/gutenberg/assets/1204802/2a3d4bbd-1ae5-4476-80b2-7483f885c373">

The revisions area should stick to the bottom closer to the save hub, even when there's scroll-space. Technically I don't know if it should be _part_ of the save dock (Jay can probably speak better to this), but I suspect we'll want to surface buttons that look almost exactly like this one in a few other sections than just the Styles section, so there could be something to it. 

<img width="349" alt="Screenshot 2023-06-01 at 10 35 04" src="https://github.com/WordPress/gutenberg/assets/1204802/da57f7cc-d306-402e-800e-0f6f48ff1a49">

We can probably get a bit closer to Jay's design here:

<img src="https://user-images.githubusercontent.com/846565/240861853-6b2dae7c-a7b3-48d6-b910-e50243fdcaf6.png">

The spacing is a bit tighter between borders, and there's no border _above_ the revisions bar.

jasmussen commented:
Almost certainly separate, and related to conversations about modals and their initial focus management very probably, but this is a good example of why setting focus on the first interactive element inside isn't that useful:

<img width="1392" alt="Screenshot 2023-06-01 at 10 35 42" src="https://github.com/WordPress/gutenberg/assets/1204802/c03cfddf-7b1c-4e3d-af52-6c1a979598db">

The close button gets way too much prominence when immediately focused like this, which isn't that useful since clicking it takes you here:

<img width="1392" alt="Screenshot 2023-06-01 at 10 35 48" src="https://github.com/WordPress/gutenberg/assets/1204802/63dcf4a3-24c6-4c17-9767-ce3f1ee31f95">

Again, not blocking for this PR, but just an example of some of the modal conversations that I believe @ciampo has been looking into.

ciampo commented:
> Again, not blocking for this PR, but just an example of some of the modal conversations that I believe @ciampo has been looking into.

Will keep it in mind when we'll start looking into refreshing the `Modal` component (right now, not as high of a priority as `DropdownMenu` and a few other components)

andrewserong committed:
Small code quality improvements

andrewserong commented:
Thanks for taking this for a spin @jasmussen!

> The revisions area should stick to the bottom closer to the save hub, even when there's scroll-space. Technically I don't know if it should be part of the save dock (Jay can probably speak better to this), but I suspect we'll want to surface buttons that look almost exactly like this one in a few other sections than just the Styles section, so there could be something to it.

Good question. For implementation in this PR, I went with adding the button to the `footer` area of the `SidebarNavigationScreen`, which was added in #50767 by @SaxonF and @ramonjd for the page detail screen. I imagine whichever direction we go with here, we might want consistency, since they both display `Last modified` text? Here's how the page detail screen currently looks:

<img width="802" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/838dae7d-930f-44a4-8c13-44bf77be3cb4">

For this PR, would it be worth us merging this in as-is and follow up separately with how we want the `footer` area to be displayed, or would folks prefer us to fix up the styling of the footer area in this PR? Since the design shows the button being directly above the save dock's top border (`SaveHub` component), I imagine the fix would mostly be about adjusting the styling of the `SidebarNavigationScreen` component. Happy to go either way 🙂

SaxonF commented:
> I went with adding the button to the footer area of the SidebarNavigationScreen

👍 This is the right approach and was the reason for adding the footer prop. 

> would folks prefer us to fix up the styling of the footer area in this PR?

It is a bit of a pain because the structure of the sidebar is a little messy but I think it's worth fixing in this PR.

andrewserong commented:
> It is a bit of a pain because the structure of the sidebar is a little messy but I think it's worth fixing in this PR.

Thanks for confirming! I'm happy to work on it in this PR 👍

So it sounds like we want to:

* Reduce the spacing within the footer area
* Stick the footer to the bottom of the sidebar in all cases, not just when there's enough content to push it right to the bottom
* Remove the top grey line?

I can chip away at that, so we've got something closer to Jay's design.

jasmussen commented:
> Reduce the spacing within the footer area
> Stick the footer to the bottom of the sidebar in all cases, not just when there's enough content to push it right to the bottom
> Remove the top grey line?

That sounds right to me, but I'll let Saxon correct me if I'm off on this.

ntsekouras commented on packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js:
```diff
@@ -74,13 +82,63 @@ function SidebarNavigationScreenGlobalStylesContent() {
+				label={ __( 'Revisions' ) }
+				onClick={ onClickRevisions }
+			>
+				<HStack
```
Technically we need [phrasing content ](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#technical_summary)elements inside `button`, so these should be `spans`. Also there are some existing `divs` inside `SidebarNavigationItem` - not introduced here.

ntsekouras reviewed (commented):
Code wise this looks good, thanks @andrewserong  !

andrewserong committed:
Try sticking the navigation screen footer to the bottom of the screen at all times

andrewserong commented:
Thanks folks! Update: I _think_ I've managed to get the sticky footer area of the navigation screen to work correctly. I've pushed an update in https://github.com/WordPress/gutenberg/pull/51149/commits/cee70819bd63425827c58d1259fda9e9e5c41689. The changes involved:

* In the sidebar, set the parent navigation screen (where the scrollbar exists) to be `flex` with `column` flex direction and height of `100%` — this allows us to grow the child content area to fill the available space, while also sticking the footer area to the bottom.
* Set the main area for the `SidebarNavigationScreen` to `flex-grow: 1` so that it expands to fill the available area.
* Move the footer area to be a sibling of that main area so that it can always stick to the bottom.
* Add a new `edit-site-sidebar-navigation-screen__footer` className for the footer area, with more limited styling (basically just the background color and padding)
* Update the page details screen so that the footer details are correctly aligned

Here's how it's looking now, along with the sticky footer behaviour:

### Styles Screen

https://github.com/WordPress/gutenberg/assets/14988353/e2a683d6-73c8-47bb-a98b-9628e2e32588

### Page Details

https://github.com/WordPress/gutenberg/assets/14988353/687076f3-d7e4-4574-8d0b-e99a1e7a75a3

I think this is ready for another round of design feedback — is this closer to the intended result? And do the changes here cause any other issues across any of the screens?

jameskoster commented:
Stickiness is working well 👍 

The dimensions are technically correct, and match other menu items. It looks a _little_ off due to the left/right margins not aligning nicely with the save dock stroke, but that's a separate issue.

The only (tiny) detail missing is the rounded corners.

andrewserong committed:
Add as span to the HStack

andrewserong committed:
Update border radius and left padding so that the button lines up

andrewserong committed:
Remove classnames call

andrewserong commented:
> The dimensions are technically correct, and match other menu items. It looks a little off due to the left/right margins not aligning nicely with the save dock stroke, but that's a separate issue.

Thanks @jameskoster, I've updated the dimensions slightly so that it lines up with the save dock stroke — I _think_ this should be the right position now, but we'll have a better idea of where it all sits once the update to the `SaveHub` lands in https://github.com/WordPress/gutenberg/pull/50567 where the stroke will then be full width.

I've also added in the border radius using `$radius-block-ui`, so here's how the button is now looking:

<img width="355" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/5d88d8b4-8ca8-4937-bf01-20af8ca98584">

This PR should be ready for a final review I think now. Unless there are any blockers, I'd probably lean toward landing this PR in pretty close to its current shape now and looking at tweaks in follow-ups, since it looks like there are a few different styling things changing across multiple PRs. That said, happy to keep chipping away here if folks would prefer to wait for other PRs to land first 🙂

ramonjd commented on packages/edit-site/src/components/global-styles/ui.js:
```diff
@@ -326,6 +344,7 @@ function GlobalStylesUI() {
 
 			<GlobalStylesActionMenu />
 			<GlobalStylesBlockLink />
+			<GlobalStylesRevisionLink />
```
I think we might need to feed some more context to this hook. It seems to prevent switching between revisions and the stylebook.


https://github.com/WordPress/gutenberg/assets/6458278/e3cbb268-8ab0-49e3-8b89-e227885113a6

My assumption is that `editorCanvasContainerView === 'global-styles-revisions'` is already `true` so the panel doesn't switch.

Maybe we need something like `canvasMode === 'edit'` or something?

andrewserong replied:
Oh, good catch! I'll look into it, but yes, hopefully something like `canvasMode === 'edit'` might work.

andrewserong replied:
I think I've figured it out @ramonjd — it seems that the `useEffect` was firing correctly (only when the container is updated to `global-styles-revisions`), but that the Style Book button didn't appear to play nicely when the path was explicitly set to `/revisions`. That is to say, opening the Style Book doesn't update the path again. As a fix, I've updated this `useEffect` so that it also handles redirects for when the canvas container is updated to `style-book`, by redirecting to the root.

I believe this is needed because the style book button lives outside of the navigation provider for global styles, so it's probably good for consistency for us to handle the redirects in the one place.

I've updated in https://github.com/WordPress/gutenberg/pull/51149/commits/51e26bc9e1ad17963f71ceea40b4afae29602013 and https://github.com/WordPress/gutenberg/pull/51149/commits/8ae9c7ab372ed677d6370bc077e37828c87e2c36 — the latter ensures that we only redirect to the root path when we're on the revisions screen. I _think_ the revisions screen is the only one where the style book won't work nicely, and this ensures that if someone is deeply nested within the global styles screens, opening the style book doesn't unexpectedly redirect back to root.

Let me know if you think this is an okay direction or if we should go in a different route... so to speak! 😄

ramonjd replied:
> Let me know if you think this is an okay direction or if we should go in a different route... so to speak!

I'm sure we can navigate towards a good solution :trollface:

andrewserong replied:
Zing! 🤣

andrewserong committed:
Add redirect for style book, too

andrewserong committed:
Only redirect to root when opening the style book on the revisions screen

ramonjd commented on packages/edit-site/src/components/global-styles/ui.js:
```diff
@@ -237,6 +237,33 @@ function GlobalStylesBlockLink() {
+			// redirect to the revisions screen.
+			goTo( '/revisions' );
+		} else if (
+			editorCanvasContainerView === 'style-book' &&
```
Thanks for working this one out! I can now switch between views.

Do you think it would be okay to keep it generic, that is, if any container view is open? I'm wondering in case another container view is added down the road.

I tested and it appears to work for this situation.

```suggestion
			!! editorCanvasContainerView &&
```

andrewserong replied:
Good call! Yes, keeping it generic will help if and when we add any other container views as this logic should redirect away from the revisions screen if the container view is anything other than revisions and we're on the revisions screen 👍

That appears to work well to me, too, so I've committed in https://github.com/WordPress/gutenberg/pull/51149/commits/e35f61b4e2ca8f3ae0916ce1487a52e0ac5094f7 with a slightly updated comment.

ramonjd reviewed (approved):
Tested in Safari/Chrome/Firefox

Thanks for working on this and coming up with a nice footer solution! 

If possible, it would be great to have a generic check for `editorCanvasContainerView` in `GlobalStylesEditorCanvasContainerLink()`... only if you think it matches the behavior you're after, otherwise LGTM.

andrewserong committed:
Use truthy check for editor canvas container when redirecting from the revisions screen.

Co-authored-by: ramon <ramonjd@gmail.com>

andrewserong commented:
> If possible, it would be great to have a generic check for editorCanvasContainerView in GlobalStylesEditorCanvasContainerLink()... only if you think it matches the behavior you're after, otherwise LGTM.

Looks good to me, thanks for the suggestion and for re-testing! I've committed that. Once the tests pass, I'll merge this in and happy to continue with tweaks in follow-ups.

jameskoster commented:
> I've updated the dimensions slightly so that it lines up with the save dock stroke

Apologies, I should have been clearer in my last comment. The revisions button had the correct dimensions, it was the stroke on the save dock that we needed to update 🙈 

No worries, I'll try to open a PR to address this.

andrewserong commented:
> The revisions button had the correct dimensions, it was the stroke on the save dock that we needed to update 🙈 No worries, I'll try to open a PR to address this.

Thanks for opening up the follow-up PR! I left a comment on it, but the additional padding was added to ensure that the button lines up correctly with the save dock's new save button dimensions. It sounds like we might need to find a balance between aligning with other elements within the content area, and aligning with the save button 🤔
